test/testdata/resolver/bind_self_type.rb

```diff
@@ -0,0 +1,75 @@
+# typed: true
+
+module Mixin
+  extend T::Sig
+
+  sig { params(block: T.proc.bind(T.self_type).void).void }
+  def foo(&block); end
+end
+
+class IncludesMixin
+  include Mixin
+
+  def does_exist; end
+
+  def bar
+    foo do
+      T.reveal_type(self) # error: Revealed type: `IncludesMixin`
+
+      does_exist
+      does_not_exist # error: Method `does_not_exist` does not exist on `IncludesMixin`
+    end
+  end
+end
+
+IncludesMixin.new.foo do
+  T.reveal_type(self) # error: Revealed type: `IncludesMixin`
+
+  does_exist
+  does_not_exist # error: Method `does_not_exist` does not exist on `IncludesMixin`
+end
+
+class Base
+  extend T::Sig
+
+  sig { params(block: T.proc.bind(T.self_type).void).void }
```

Comment:
Could you make `Base` generic, and show that it's possible to rebind to one of its type members? That would make a good negative test for the current behavior in type_syntax.cc.

Also revealing the type of `self` inside of a block would help illustrate the problem here.